### PR TITLE
Added authentication middleware

### PR DESCRIPTION
**Pull request by yoannpicquenot**

### CODE REVIEW

File: middleware/index.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+'use strict';
```
Comment (Quentin01) [inline middleware/index.js]: Why are you naming this middleware `index.js`?

File: middleware/index.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+var config = require('../config');
+
+module.exports = (req, res, next) => {
+  let accessToken = req.headers['x-access-token'];
```
Comment (Quentin01) [inline middleware/index.js]: Don't we use 4 spaces for indentation everywhere in the project? (We can change that maybe)

File: middleware/index.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+  if(!accessToken) {
+    res.status(401).send('Authentication token is required');
+  } else if (accessToken !== config.auth_token) {
+    res.status(401).send('Authentication token is not correct');
+  }
+  next();
```
Comment (Quentin01) [inline middleware/index.js]: I think that you don't want to call next in case of error 401

File: server.js [hunk 4]
```diff
@@ -9,6 +9,7 @@ var config = require('./config');
 var server = express();
 
 var api = require('./index');
+var middleware = require('./middleware');
```
Comment (Quentin01) [inline server.js]: Change also the name of this variable, it's an authentification middleware

File: server.js [hunk 5]
```diff
@@ -30,6 +31,7 @@ mongoose.connect(config.mongodb_uri, (err) => {
     server.use(morgan(':method :url | :status | :response-time ms'));
     server.use(express.static(__dirname));
     server.use(bodyparser.json({limit: '50mb'}));
+    server.use('/api/', middleware);
```
Comment (Quentin01) [inline server.js]: Why are you protecting `/api/` with the authentification middleware?

**Review by Quentin01 — CHANGES_REQUESTED** [review]

File: middleware/authentication.js [hunk 6]
```diff
@@ -0,0 +1,18 @@
+'use strict';
+
+var config = require('../config');
+
+const methodToCatch = ['PUT', 'DELETE', 'POST'];
+
+module.exports = (req, res, next) => {
+    if (methodToCatch.indexOf(req.method) !== -1) {
```
Comment (saveman71) [inline middleware/authentication.js]: This is very very basic but it's a start. What if it's just a GET? No next?

File: server.js [hunk 7]
```diff
@@ -9,6 +9,7 @@ var config = require('./config');
 var server = express();
 
 var api = require('./index');
+var middlewareAuthentication = require('./middleware/authentication.js');
```
Comment (saveman71) [inline server.js]: -.js

File: config/index.js [hunk 8]
```diff
@@ -2,5 +2,6 @@ module.exports = {
     server: {
         port: process.env.PORT || 5000
     },
-    mongodb_uri: process.env.MONGODB_URI
+    mongodb_uri: process.env.MONGODB_URI,
+    auth_token: process.env.AUTH_TOKEN
```
Comment (saveman71) [inline config/index.js]: Maybe you should add a check earlier in the file to ensure the auth_token is an ok value, or later you'll check against `undefined`

**Review by saveman71 — CHANGES_REQUESTED** [review]

File: middleware/authentication.js (hunk 6 above)
Comment (Quentin01) [inline middleware/authentication.js]: Why do you want to only catch these methods? We will maybe have some `GET` endpoints that we will protect.

Comment (Quentin01) [inline middleware/authentication.js]: You have still failed your conditions, `next()` is not called when the access token is good.

File: server.js [hunk 9]
```diff
@@ -9,34 +9,25 @@ var config = require('./config');
...
+    }
 
+    server.use(middlewareCORS);
     server.use(morgan(':method :url | :status | :response-time ms'));
     server.use(express.static(__dirname));
-    server.use(bodyparser.json({limit: '50mb'}));
+    server.use(bodyparser.json({
+        limit: '50mb'
+    }));
+    server.use('/api/', middlewareAuthentication);
```
Comment (Quentin01) [inline server.js]: Why are you protecting all the endpoints? We just want to protect the sensitive endpoints (admin ones)

**Review by Quentin01 — CHANGES_REQUESTED** [review]

**Comment by yoannpicquenot** [comment]
@Quentin01 can you check please?

**Review by Quentin01 — APPROVED** [review]

**Comment by Quentin01** [comment]
@saveman71 Do you also approve the final changes?

**Review by saveman71 — APPROVED** [review]

**Pull request merged**